tests/framework/extensions/nodes/nodes.go

```diff
@@ -0,0 +1,63 @@
+package nodes
+
+import (
+	"context"
+
+	"github.com/rancher/rancher/pkg/api/scheme"
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var NodeGroupVersionResource = schema.GroupVersionResource{
+	Group:    "",
+	Version:  "v1",
+	Resource: "nodes",
+}
+
+// get the nodes
+func GetNodes(client *rancher.Client, clusterID string, listOpts metav1.ListOptions) ([]corev1.Node, []error) {
+
+	var nodesList []corev1.Node
+	dynamicClient, err := client.GetDownStreamClusterClient(clusterID)
+	if err != nil {
+		return nil, []error{err}
+	}
+	nodeResource := dynamicClient.Resource(NodeGroupVersionResource)
+	nodes, err := nodeResource.List(context.TODO(), listOpts)
+	if err != nil {
+		return nil, []error{err}
+	}
+	for _, unstructuredNode := range nodes.Items {
+		newNode := &corev1.Node{}
+		err := scheme.Scheme.Convert(&unstructuredNode, newNode, unstructuredNode.GroupVersionKind())
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		nodesList = append(nodesList, *newNode)
+	}
+	return nodesList, []error{err}
+}
+
+// get node IP, user needs to pass which type they want ExternalIP, InternalIP, Hostname, check core/v1/types.go
+func GetNodeIP(nodes []corev1.Node, nodeAddressType string) ([]string, []error) {
+	var nodeErrors []error
+	var nodeIPList []string
+	var nodeIP string
+
+	for _, node := range nodes {
+		nodeAddressList := node.Status.Addresses
+		for _, ip := range nodeAddressList {
+			if ip.Type == corev1.NodeAddressType(nodeAddressType) {
+				nodeIP = ip.Address
```

Comment:
is there more than 1 type per node.Status? i.e. can this have more than 1 externalIP? If not, you could break the loop here in the if statement.

Reply:
So there are 5 types. And I see, I'll add to break after the condition is met